File: crates/godwinmix-client/src/frames.rs
//! Multiview frames off the wire.
//!
//! A binary frame on `/rpc` is a 16 byte header then JPEG:
//!
//! ```text
//! offset 0   u32  seq                little endian
//! offset 4   u32  layout id          little endian
//! offset 8   u64  running time ms    little endian
//! offset 16  ...  JPEG bytes
//! ```
//!
//! The layout id matches the id in `event/multiview.layout`, which is what
//! lets a client cut cells out of a sheet without a race when the layout
//! changes mid flight. The core's writer is `api::rpc::frame_header`, and the
//! test at the bottom of this file checks this reader against the bytes that
//! writer produces.

pub const HEADER_BYTES: usize = 16;

/// One mosaic frame, header read and JPEG still compressed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Frame {
    /// Counts up per connection. Wraps rather than stopping.
    pub seq: u32,
    /// Which `event/multiview.layout` this picture was cut for.
    pub layout: u32,
    /// Programme running time the frame was taken at, in milliseconds.
    pub running_time_ms: u64,
    pub jpeg: Vec<u8>,
}

impl Frame {
    /// The cell rectangle for one source, from the layout with this frame's id.
    ///
    /// Returns None when the layout is a different one, which is the case a
    /// client hits for a frame or two after the grid changes.
    pub fn cell<'a>(
        &self,
        layout: &'a crate::MultiviewLayout,
        source: &str,
    ) -> Option<&'a crate::CellAssignment> {
        if layout.id != self.layout {
            return None;
        }
        layout.cells.iter().find(|c| c.source.as_deref() == Some(source))
    }
}

/// Split one binary message into its header and its JPEG.
///
/// Returns None for anything too short to be a frame, which is what a client
/// gets if it points at a core that sends bare JPEGs.
pub fn parse_frame(bytes: &[u8]) -> Option<Frame> {
    if bytes.len() <= HEADER_BYTES {
        return None;
    }
    let seq = u32::from_le_bytes(bytes[0..4].try_into().ok()?);
    let layout = u32::from_le_bytes(bytes[4..8].try_into().ok()?);
    let running_time_ms = u64::from_le_bytes(bytes[8..16].try_into().ok()?);
    Some(Frame { seq, layout, running_time_ms, jpeg: bytes[HEADER_BYTES..].to_vec() })
}

/// The mosaic width to ask the core for.
///
/// The sheet holds `cols` cells across. A tile that is `tile_px` real pixels
/// wide wants `tile_px * cols`, rounded up to a multiple of 16 because
/// encoders like even macroblocks, and clamped to what the protocol allows.
/// Asking for more than this is bytes nobody looks at.
pub fn sheet_width_for(tile_px: u32, cols: u32) -> u32 {
    let sheet = tile_px.max(1) * cols.max(1);
    let rounded = sheet.div_ceil(16) * 16;
    rounded.clamp(320, 1920)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The same bytes `api::rpc::frame_header` writes, built here by hand so
    /// this is a test of the pair and not a restatement of one side.
    fn header(seq: u32, layout: u32, ms: u64) -> Vec<u8> {
        let mut out = Vec::with_capacity(HEADER_BYTES);
        out.extend_from_slice(&seq.to_le_bytes());
        out.extend_from_slice(&layout.to_le_bytes());
        out.extend_from_slice(&ms.to_le_bytes());
        out
    }

    #[test]
    fn reads_a_frame() {
        let mut bytes = header(7, 3, 1234);
        bytes.extend_from_slice(&[0xff, 0xd8, 0xff, 0xe0]);
        let frame = parse_frame(&bytes).expect("a 20 byte message is a frame");
        assert_eq!(frame.seq, 7);
        assert_eq!(frame.layout, 3);
        assert_eq!(frame.running_time_ms, 1234);
        assert_eq!(frame.jpeg, vec![0xff, 0xd8, 0xff, 0xe0]);
    }

    #[test]
    fn a_header_with_no_picture_is_not_a_frame() {
        assert!(parse_frame(&header(1, 1, 1)).is_none());
        assert!(parse_frame(&[]).is_none());
    }

    #[test]
    fn widths_are_rounded_and_clamped() {
        assert_eq!(sheet_width_for(100, 3), 304_u32.max(320));
        assert_eq!(sheet_width_for(320, 3), 960);
        assert_eq!(sheet_width_for(1000, 4), 1920);
    }
}

File: ui/client/frames.js
// Multiview frames off the wire and onto tiles.
//
// A frame from `/rpc` is a 16 byte header then JPEG:
//
//   offset 0   u32  seq              little endian
//   offset 4   u32  layout id        little endian
//   offset 8   u64  running time ns  little endian
//   offset 16  ...  JPEG bytes
//
// The layout id matches the id in `event/multiview.layout`, which is what lets
// a client cut cells out of a sheet without a race when the layout changes mid
// flight. Today's server sends a bare JPEG with no header; the legacy adapter
// hands frames here with `bare: true` and the layout the status document
// carries.
//
// One socket carries two pictures: the mosaic, and the armed scene for the
// pane beside the programme. The top bit of the sequence number says which,
// because the header had no spare field and every client reads it at sixteen
// bytes. A preview frame is one whole picture and names no grid.

export const HEADER_BYTES = 16;
export const PREVIEW_STREAM = 0x80000000;

/** Split one binary frame into its header and its JPEG. */
export function parseFrame(buffer) {
  const bytes = buffer instanceof ArrayBuffer ? new Uint8Array(buffer) : buffer;
  if (bytes.length <= HEADER_BYTES) return null;
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const seq = view.getUint32(0, true);
  return {
    seq: seq & ~PREVIEW_STREAM,
    preview: (seq & PREVIEW_STREAM) !== 0,
    layout: view.getUint32(4, true),
    runningTimeMs: view.getBigUint64(8, true),
    jpeg: bytes.subarray(HEADER_BYTES),
  };
}

/** Wrap a bare JPEG so callers see one shape whichever transport delivered it. */
export function bareFrame(buffer, layout) {
  const bytes = buffer instanceof ArrayBuffer ? new Uint8Array(buffer) : buffer;
  // The legacy stream is the mosaic and only ever was.
  return { seq: 0, preview: false, layout: layout || 0, runningTimeMs: 0n, jpeg: bytes };
}

/**
 * Holds the newest sheet as one ImageBitmap and paints cells out of it.
 *
 * One decode per frame for the whole mosaic, then a canvas blit per visible
 * tile. Decoding once and blitting many is the difference between a Pi keeping
 * up and not. `createImageBitmap` is present in WebView2, WebKitGTK and Safari
 * 15 and later, so all three shells take the same path.
 */
export class SheetPainter {
  constructor() {
    this.bitmap = null;
    this.layout = null;
    this.pending = false;
    this.targets = new Map(); // canvas -> cell index
    this.blobUrl = null;
  }

  /** The cell table from `event/multiview.layout` or from status.multiview. */
  setLayout(layout) {
    this.layout = layout;
  }

  /** Register a canvas to be painted with one cell. Returns a release function. */
  attach(canvas, cellIndex) {
    this.targets.set(canvas, cellIndex);
    this._paint();
    return () => this.targets.delete(canvas);
  }

  get wanted() {
    return this.targets.size > 0;
  }

  /** Take a frame. Decoding is async, so late frames are dropped, not queued. */
  async push(frame) {
    if (!this.wanted || this.pending) return;
    this.pending = true;
    try {
      const blob = new Blob([frame.jpeg], { type: "image/jpeg" });
      const decoded = await decode(blob);
      if (this.bitmap && this.bitmap.close) this.bitmap.close();
      this.bitmap = decoded;
      this._paint();
    } catch {
      // A truncated frame is not worth a message: the next one is along.
    } finally {
      this.pending = false;
    }
  }

  _paint() {
    if (!this.bitmap || !this.layout) return;
    const cells = this.layout.cells || [];
    for (const [canvas, index] of this.targets) {
      const cell = cells.find((c) => c.index === index);
      if (!cell) continue;
      const w = canvas.width;
      const h = canvas.height;
      if (!w || !h) continue;
      const ctx = canvas.getContext("2d", { alpha: false });
      if (!ctx) continue;
      ctx.drawImage(this.bitmap, cell.x, cell.y, cell.w, cell.h, 0, 0, w, h);
    }
  }

  destroy() {
    if (this.bitmap && this.bitmap.close) this.bitmap.close();
    this.bitmap = null;
    this.targets.clear();
  }
}

/**
 * The newest whole picture, painted onto every canvas attached to it.
 *
 * The preview is one frame rather than a sheet, so there is no layout to wait
 * for and nothing to cut out: decode, then blit. Same shape as `SheetPainter`
 * so a panel holds either the same way.
 */
export class PicturePainter {
  constructor() {
    this.bitmap = null;
    this.pending = false;
    this.targets = new Set();
  }

  /** Register a canvas to be painted with the whole picture. */
  attach(canvas) {
    this.targets.add(canvas);
    this._paint();
    return () => this.targets.delete(canvas);
  }

  get wanted() {
    return this.targets.size > 0;
  }

  /** Take a frame. Late frames are dropped rather than queued, as the sheet does. */
  async push(frame) {
    if (!this.wanted || this.pending) return;
    this.pending = true;
    try {
      const decoded = await decode(new Blob([frame.jpeg], { type: "image/jpeg" }));
      if (this.bitmap && this.bitmap.close) this.bitmap.close();
      this.bitmap = decoded;
      this._paint();
    } catch {
      // A truncated frame is not worth a message: the next one is along.
    } finally {
      this.pending = false;
    }
  }

  _paint() {
    if (!this.bitmap) return;
    for (const canvas of this.targets) {
      if (!canvas.width || !canvas.height) continue;
      const ctx = canvas.getContext("2d", { alpha: false });
      if (ctx) ctx.drawImage(this.bitmap, 0, 0, canvas.width, canvas.height);
    }
  }

  destroy() {
    if (this.bitmap && this.bitmap.close) this.bitmap.close();
    this.bitmap = null;
    this.targets.clear();
  }
}

/**
 * One JPEG to something `drawImage` accepts.
 *
 * `createImageBitmap` is the one to use: it decodes off the main thread and the
 * result can be closed, so a wall of tiles at eight frames a second does not
 * leave a hundred decoded pictures for the collector. WebKit before Safari 15
 * has no such thing, and the Tauri window on an older macOS is exactly that,
 * so an <img> over an object URL stands in. Same interface to the caller.
 */
function decode(blob) {
  if (typeof createImageBitmap === "function") return createImageBitmap(blob);
  return new Promise((resolve, reject) => {
    const url = URL.createObjectURL(blob);
    const img = new Image();
    img.onload = () => {
      // Revoking inside onload, never before: revoking early blanks the image
      // on some builds and leaks one URL per frame on the rest.
      URL.revokeObjectURL(url);
      resolve(img);
    };
    img.onerror = () => {
      URL.revokeObjectURL(url);
      reject(new Error("the frame did not decode"));
    };
    img.src = url;
  });
}

/**
 * The width to ask the core for.
 *
 * The mosaic is one sheet holding `cols` cells across. A tile is `cssWidth`
 * device independent pixels wide on a screen with `devicePixelRatio` real
 * pixels per one of those. Asking for more than that is bytes nobody sees;
 * asking for less is a soft picture. Rounded up to a multiple of 16 because
 * encoders like even macroblocks, and clamped to the range the protocol allows.
 */
export function sheetWidthFor(cssWidth, cols, dpr) {
  const ratio = dpr || (typeof devicePixelRatio === "number" ? devicePixelRatio : 1);
  const perCell = Math.ceil(cssWidth * ratio);
  const sheet = perCell * Math.max(1, cols || 1);
  const rounded = Math.ceil(sheet / 16) * 16;
  return Math.max(320, Math.min(1920, rounded));
}
